### ISSUE TITLE
Support cwd of current package inside monorepo

### ISSUE THREAD

**Issue by cdauth** [firsttris/vscode-jest-runner#353]
I am using this plugin in a monorepo with several modules. In my case, I have configured `jestrunner.jestCommand` to be `yarn test`, which in my case executes `vitest` instead of `jest`.

By default, vscode-jest-runner executes the command on the root level of my monorepo. I would like it to execute on the root level of the individual NPM package instead, for several reasons:
* I don't have `jest` (or in my case `vitest`) installed on the root level
* The implementation of my `yarn test` command might be different for each module

As a workaround, I currently have `jestrunner.changeDirectoryToWorkspaceRoot` set to `false` and manually `cd` into the appropriate directory in the `jest` terminal before running the tests.

One way this could be achieved would be to add a config option (such as `jestrunner.changeDirectoryToPackageRoot`) that `cd`s to the nearest ancestor directory containing a `package.json` file before running a test.

On the other hand, I'm wondering whether this behaviour shouldn't be the default. Are there any setups that this would break?

**Comment by bryceg** [comment]
similar issue here with monorepo, its cwd is down in the root.  An option to specify the directory through a configuration option so I can use something like a vscode variable reference would help.

**Comment by firsttris** [comment]
you are right, this is a missing feature, some people try to implement it, but this lead to a breaking change for other project.

i think i was this PR: https://github.com/firsttris/vscode-jest-runner/issues/220

Still waiting for someone contribute this option.

Most people use NX for Monorepo, NX installs this extension as recommended extension.
NX has single version policy and all dependencies are in root.

best regards
Tristan

**Comment by domsleee** [comment]
From `jestRunnerConfig.ts` when it resolves `cwd()`, it looks for the nearest package, but only if the package has `node_modules/jest`:

https://github.com/firsttris/vscode-jest-runner/blob/81b797c77eb23fbc5c07a8a953a57b15edf0685a/src/jestRunnerConfig.ts#L64-L74

So the current fallback logic for `jestrunner.changeDirectoryToWorkspaceRoot` is:
1. `jestrunner.projectPath`
2. (code snippet above) the nearest `package.json` with jest in it's node_modules folder
3. `${workspaceFolder}`

Would it make sense to change (2) to find packages with either `jest` or `vitest` in it's node_modules?

**Comment by cdauth** [comment]
Adding support for `vitest` might be nice, but it does not solve the problem that I describe here.

The problem described here is that dependency hoisting causes jest/vitest to be installed in a different location than where it should be used. For example, the directory layout might look like this:
```
/
/package.json
/node_modules/.bin/jest
/module1/package.json
/module2/package.json
```

jest is only listed as a dependency in `/module1/package.json` and `/module2/package.json` and `test` scripts are defined in those files. In `/package.json`, jest is not defined as a dependency and there is no `test` script. But vscode-jest-runner would run the tests in `/`.

If `module1` and `module2` would for some reason depend on different versions of jest, the directory layout would change into this:
```
/
/package.json
/module1/package.json
/module1/node_modules/.bin/jest
/module2/package.json
/module2/node_modules/.bin/jest
```

So vscode-jest-runner would now run the tests inside the respective module folders.

Personally I don't see a use case where in the first case (both modules using the same jest version) the tests should run on the root level of the project. So my recommendation would be to simply remove the check for `node_modules/.bin/jest` altogether and simply run the tests in the nearest directory that contains a `package.json`. But maybe there is some use case that I'm not aware of.

**Comment by digitalmaster** [comment]
This is my exact issue as well. Also, agree that it seems like this would work as the default. But would be happy with an option.

**Comment by ghost91-** [comment]
but `jest` will not be in the `node_modules` of the current package all the time. It might be hoisted up to the root `node_modules` folder instead. It would be nice to have a way to force `vscode-jest-runner ` to simply change to the closest directory that has a `package.json`, regardless of anything else. That solve the problem for my monorepo setup.

**Comment by dobesv** [comment]
Also consider for yarn users that jest might not be in `node_modules` at all but in some completely different location.  Is the `node_modules` check actually helpful?  Why not just take the nearest `package.json` file without checking for anything in `node_modules` ?

**Comment by dobesv** [comment]
Something like this? https://github.com/firsttris/vscode-jest-runner/pull/384
